File: go_backend/middlewares/recovery.go
package middlewares

import (
	"fmt"
	"net/http"
	"runtime/debug"

	"syncflow/logger"
	"syncflow/services/alerts"
	"syncflow/utils"
	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

func Recovery() gin.HandlerFunc {
	return func(c *gin.Context) {
		defer func() {
			if err := recover(); err != nil {
				logger.Error("Recovery from panic",
					zap.Any("error", err),
					zap.String("stack", string(debug.Stack())),
				)

				// Send alert
				alerts.Error(fmt.Sprintf("Panic recovered: %v\nStack: %s", err, string(debug.Stack())))

				utils.Error(c, http.StatusInternalServerError, "Internal Server Error")
			}
		}()
		c.Next()
	}
}

File: node_backend/src/middlewares/recovery.ts
// middlewares/recovery.ts — Equivalent to go_backend/middlewares/recovery.go
import { Request, Response, NextFunction } from 'express';
import logger from '../logger/logger';
import * as alerts from '../services/alerts/alerts';
import * as apiResponse from '../utils/response';

/**
 * Recovery (error handler) middleware — catches unhandled errors and panics.
 * Mirrors the Go Recovery middleware.
 *
 * NOTE: In Express, this must be registered AFTER all routes (error-handling middleware).
 * Express error-handling middleware has 4 parameters (err, req, res, next).
 */
export function recoveryMiddleware(
    err: Error,
    req: Request,
    res: Response,
    _next: NextFunction,
): void {
    const stack = err.stack || '';

    logger.error('Recovery from panic', {
        error: err.message,
        stack,
    });

    // Send alert
    alerts.error(`Panic recovered: ${err.message}\nStack: ${stack}`);

    apiResponse.error(res, 500, 'Internal Server Error');
}
